llvm/lib/Support/YAMLParser.cpp

```diff
@@ -1848,13 +1869,15 @@ bool Scanner::fetchMoreTokens() {
   if (*Current == ',')
     return scanFlowEntry();
 
-  if (*Current == '-' && isBlankOrBreak(Current + 1))
+  if (*Current == '-' && (isBlankOrBreak(Current + 1) || Current + 1 == End))
     return scanBlockEntry();
 
-  if (*Current == '?' && (FlowLevel || isBlankOrBreak(Current + 1)))
+  if (*Current == '?' && (Current + 1 == End || isBlankOrBreak(Current + 1)))
     return scanKey();
 
-  if (*Current == ':' && (FlowLevel || isBlankOrBreak(Current + 1)))
+  if (*Current == ':' && ((FlowLevel && (IsAdjacentValueAllowed ||
```

Comment:
Isn't checking `FlowLevel` here redundant assuming `IsAdjacentValueAllowed` is set correctly?

I'm also confused about the condition `!isPlainSafeNonBlank`. Shouldn't we always parse a (possibly empty) value following the `:` when an adjacent value is allowed?

I also think the order should make the "normal" case of a blank/break following the indicator come first; all together is the following still correct, or am I missing something?

```
if (*Current == ':' && (isBlankOrBreak(Current + 1) || IsAdjacentValueAllowed))
```